pymtl3/passes/PassConfigs.py

```diff
@@ -0,0 +1,63 @@
+#=========================================================================
+# PassConfigs.py
+#=========================================================================
+# Author : Peitian Pan
+# Date   : Jul 26, 2019
+"""Base class of customized pass configuration classes."""
+
+from copy import copy
+
+from .errors import InvalidPassOption, InvalidPassOptionValue
+
+
+class BasePassConfigs:
+  """Base class of customized pass configrations.
+
+  Users who wish to use this Base class should (1) put the name of their
+  pass to `PassName`; (2) fill their own options and default values into
+  `Options` class attribute; (3) define `check_<option_name>` methods
+  that takes the value of `<option_name>` and returns whether this
+  option is valid.
+  """
+  def __init__( s, **kwargs ):
+    # s.Options = {}
+    # s.PassName = "<base_pass>"
+    s.options = s._gen_options( kwargs )
+
+  def check_options( s ):
+    """Return whether the given options are all valid."""
+    for opt, value in s.options.items():
+      func = getattr( s, "check_"+opt, lambda self, val: True )
+      # check_* methods will throw exceptions if the check fails
+      func(s, value)
+    return True
+
+  def get_option( s, opt ):
+    """Return the value of option `opt`."""
+    return s.options[opt]
+
+  def set_option( s, opt, value ):
+    s.options[opt] = value
+
+  def _gen_options( s, _opts ):
+    opts = copy( s.Options )
```

Comment:
Is this s.Options still a valid variable after you comment it out?

Reply:
None of the methods of the base class is supposed to be called directly. `s.Options` is an attribute that derived classes are supposed to override. I think making `Options` an abstract property might help this by raising an exception every time an abstract `Options` is used.